### PR TITLE
Add methods to set the thread affinity

### PR DESCRIPTION
**Pull request by saikishor**
This PR uses the commits from PR: #140 and add more fixes on top of it and some refactoring and renaming of methods

Closes  #140

### CODE REVIEW

**Comment by christophfroehlich** [comment]
:eyes: @firesurfer do you support the changes?
@SENAI-GilmarCorreia does this work for you on your windows setup?

**Review by christophfroehlich — COMMENTED** [review]
👀 @firesurfer do you support the changes?
@SENAI-GilmarCorreia does this work for you on your windows setup?

**Comment by SENAI-GilmarCorreia** [comment]
@christophfroehlich, I'm having trouble with the #warning on line 35 in thread_priority.hpp. I'm submitting a PR.

**Comment by saikishor** [comment]
> @christophfroehlich, I'm having trouble with the #warning on line 35 in thread_priority.hpp. I'm submitting a PR.
> 

@SENAI-GilmarCorreia the issue with the thread_priority.hpp is with this PR? Or with master branch?

**Comment by firesurfer** [comment]
I think the `message` parameter is a bit weird API wise. 

As an alternative one could either:

a) Return a std::pair<bool, std::string>
b) Make passing the message parameter optional. I am not sure if in C++17 something like this actually works:

`bool set_cpu_affinity(int pid,int core, std::string& message = {}` 

In addition I would make the `pid` parameter optional or provide an additional function: `set_current_thread_affinity`

**Comment by saikishor** [comment]
> I think the `message` parameter is a bit weird API wise. 
> 
> As an alternative one could either:
> 
> a) Return a std::pair<bool, std::string>
> b) Make passing the message parameter optional. I am not sure if in C++17 something like this actually works:
> 
> `bool set_cpu_affinity(int pid,int core, std::string& message = {}` 
> 
> In addition I would make the `pid` parameter optional or provide an additional function: `set_current_thread_affinity`
> 
> Other than that: Looks fine to me.

@firesurfer first of all, thank you for all your contributions 

Sure, I initially thought of the pair option but I skipped. The message parameter is nice, if you want to have the logging at a higher level, instead of directing it to the std::cerr. 

I can add an additional method called `set_current_thread_affinity`, it makes sense.

**Comment by saikishor** [comment]
@firesurfer can you take a look now?. If it looks good to you, please approve the PR

File: src/realtime_helpers.cpp [hunk 1]
```diff
@@ -34,10 +34,13 @@
 #include <sched.h>
 #include <sys/capability.h>
 #include <sys/mman.h>
+
+#include <unistd.h>
 #endif
 
 #include <cstring>
 #include <fstream>
+#include <iostream>
```
Comment (firesurfer) [inline src/realtime_helpers.cpp:43]: I think we do not need this include

Comment (saikishor) [reply src/realtime_helpers.cpp]: Done!

File: src/realtime_helpers.cpp [hunk 2]
```diff
@@ -112,4 +115,86 @@ bool lock_memory(std::string & message)
...
+  CPU_ZERO(&cpuset);
+  // Obtain available processors
+  const auto number_of_cores = get_number_of_available_processors();
+
+  // Reset affinity by setting it to all cores
+  if (core < 0) {
+    for (auto i = 0; i < number_of_cores; i++) {
+      CPU_SET(i, &cpuset);
+    }
+    // And actually tell the schedular to set the affinity of the currently calling thread
```
Comment (firesurfer) [inline src/realtime_helpers.cpp:160]: The comment should be updated

Comment (saikishor) [reply src/realtime_helpers.cpp]: Done! Updated

**Review by firesurfer — APPROVED** [review]

**Comment by SENAI-GilmarCorreia** [comment]
> @SENAI-GilmarCorreia the issue with the thread_priority.hpp is with this PR? Or with master branch?

@saikishor You are right! I will open a PR directly to the master branch.

**Review by SENAI-GilmarCorreia — APPROVED** [review]

**Review by christophfroehlich — APPROVED** [review]

**Pull request merged**